vulnerabilities/unbounded-return-data-gas-griefing.md

```diff
@@ -0,0 +1,87 @@
+## Unbounded Return Data Gas Griefing
+
+The [Byzantium](https://blog.ethereum.org/2017/10/12/byzantium-hf-announcement) 2017 mainnet hard-fork included [EIP-211](https://eips.ethereum.org/EIPS/eip-211), which added 2 new opcodes `RETURNDATASIZE` and `RETURNDATACOPY`. This allows for an arbitrary-length return data buffer, from which the caller can copy all or some of the return data from an external call. This variable length buffer is created empty for each new call-frame. Prior to this EIP, the return data size had to be anticipated in advance in the call parameters.
+
+However under Solidity's implementation, up until at least `0.8.26`, the entirety of this return data is automatically copied from the buffer into memory, even when using a Solidity low-level call and with the omission of the `bytes memory data` syntax.
+
+Consider the following example:
+
+```solidity
+pragma solidity 0.8.26;
+
+contract Attacker {
+    function returnExcessData() external pure returns (string memory) {
+        revert("Passing in excess data that the Solidity compiler will automatically copy to memory");   // Both statements can return unbounded data
+        return "Passing in excess data that the Solidity compiler will automatically copy to memory";
+    }
+}
+
+
+contract Victim {
+    function callAttacker(address attacker) external returns (bool) {
+        (bool success, ) = attacker.call{gas: 2500}(abi.encodeWithSignature("returnExcessData()"));
+        return success;
+    }
+}
+```
+
+In the above example one can observe that even though the `Victim` contract has not explicitly requested `bytes memory data` to be returned, and has furthermore given the external call a gas stipend of 2500, Solidity will still invoke `RETURNDATACOPY` during the top-level call-frame. This means the `Attacker` contract, through revert or return, can force the `Victim` contract to consume unbounded gas during their own call-frame and not that of the `Attacker`. Given that memory gas costs grow exponentially after 23 words, this attack vector has the potential to prevent certain contract flows from being executed.
```

Comment:
"Given that memory gas costs grow exponentially after 23 words, this attack vector has the potential to prevent certain contract flows from being executed."

Would be great to note what kind of contract flows can be prevented here, i.e. what are some attacks that are possible? Does this just work to cause user tx's to revert with out of gas?

Reply:
Yes causes 'out of gas' revert; added a sentence on undelegating/unstaking example flows